src/note.rs

```diff
@@ -0,0 +1,394 @@
+
+        buf.read_exact(&mut one_stealth_address)?;
+        n += one_stealth_address.len();
+        self.stealth_address =
+            StealthAddress::from_bytes(&one_stealth_address)?;
+
+        buf.read_exact(&mut one_u64)?;
+        n += one_u64.len();
+        self.pos = u64::from_le_bytes(one_u64);
+
+        buf.read_exact(&mut self.encrypted_data)?;
+        n += self.encrypted_data.len();
+
+        Ok(n)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+impl Note {
+    /// Creates a new phoenix output note
+    pub fn new(note_type: NoteType, psk: &PublicSpendKey, value: u64) -> Self {
+        let r = JubJubScalar::random(&mut rand::thread_rng());
+        let nonce = JubJubScalar::random(&mut rand::thread_rng());
+        let blinding_factor = JubJubScalar::random(&mut rand::thread_rng());
+
+        Self::deterministic(note_type, &r, nonce, psk, value, blinding_factor)
+    }
+
+    /// Creates a new transparent note
+    pub fn transparent(psk: &PublicSpendKey, value: u64) -> Self {
+        Self::new(NoteType::Transparent, psk, value)
+    }
+
+    /// Creates a new obfuscated note
+    pub fn obfuscated(psk: &PublicSpendKey, value: u64) -> Self {
+        Self::new(NoteType::Obfuscated, psk, value)
+    }
+
+    /// Create a new phoenix output note without inner randomness
+    pub fn deterministic(
+        note_type: NoteType,
+        r: &JubJubScalar,
+        nonce: JubJubScalar,
+        psk: &PublicSpendKey,
+        value: u64,
+        blinding_factor: JubJubScalar,
+    ) -> Self {
+        let stealth_address = psk.gen_stealth_address(r);
+
+        let value_commitment = JubJubScalar::from(value);
+        let value_commitment = JubJubExtended::from(JubJubAffine::from(
+            &(GENERATOR_EXTENDED * value_commitment)
+                + &(GENERATOR_NUMS_EXTENDED * blinding_factor),
+        ));
+
+        // Output notes have undefined position
+        let pos = 0;
+
+        let encrypted_data = match note_type {
+            NoteType::Transparent => {
+                chunk_of!(ENCRYPTED_DATA_SIZE; &value.to_le_bytes())
+            }
+            NoteType::Obfuscated => {
+                let shared_secret = dhke(r, psk.A());
+                let nonce = BlsScalar::from(nonce);
+                let value = BlsScalar::from(value);
+                let blinding_factor = BlsScalar::from(blinding_factor);
+
+                let mut cipher = PoseidonCipher::encrypt(
+                    &[value, blinding_factor],
+                    &shared_secret,
+                    &nonce,
+                );
+
+                // TODO: replace the internals of PoseidonCipher with a plain
+                // array.
+                // Adding an infallible `to_bytes` method too.
+                let mut encrypted_data = [0u8; 96];
+                cipher
+                    .read_exact(&mut encrypted_data)
+                    .expect("Cannot encode PoseidonCipher");
```

Comment:
If the operation is fallible then we should work with `Result`. Since the `Note` contains other scalars, then we have no strict requirement to use only bytes.

Reply:
In order to use directly the `PoseidonCipher` as `encrypted_data`, that was my initial idea, we need few small changes in the `poseidon` repo, otherwise we might complicate a bit the code here.

Since `encrypted_data` is encapsulated and not exposed, I don't think it's a blocker for the landing. I will therefore proceed to merge this PR, open a PR for the `poseidon252` repo, and once the changes on the `poseidon252` are landed proceed to change the internals in the `Note`.